scripts/tests/twister/test_handlers.py

```diff
@@ -0,0 +1,822 @@
+            return mock.Mock(), mock.Mock()
+
+    def assert_popen(command, stdout=None, stderr=None, cwd=None, env=None, shell=None):
+        return mock.Mock(
+            __enter__=mock.Mock(return_value=mock_process),
+            __exit__=mock.Mock(return_value=None)
+        )
+
+    mock_process.communicate = mock.Mock(side_effect=raise_timeout_fn)
+
+    script = [os.path.join('test','script', 'path'), 'arg']
+    timeout = 60
+
+    with mock.patch("subprocess.Popen", mock.Mock(side_effect=assert_popen)):
+        DeviceHandler.run_custom_script(script, timeout)
+
+    if raise_timeout:
+        assert 'timeout' in caplog.text.lower() or 'timed out' in caplog.text.lower()
+        mock_process.assert_has_calls([mock.call.communicate(timeout=timeout), mock.call.kill(), mock.call.communicate()])
+    elif mock_process.returncode == 0:
+        assert not any([r.levelname == 'ERROR' for r in caplog.records])
+    else:
+        assert 'custom script failure' in caplog.text.lower()
+
+
+TESTDATA_7 = [
+    (0, False),
+    (4, False),
+#  (0, True) # Known bug
+]
+
+@pytest.mark.parametrize("num_of_failures, raise_exception", TESTDATA_7)
+def test_devicehandler_get_hardware(mocked_instance, caplog, num_of_failures, raise_exception):
+    expected_hardware = mock.Mock()
+
+    def mock_availability(handler, instance, no=num_of_failures):
+        if raise_exception:
+            raise TwisterException(f'dummy message')
+        if handler.no:
+            handler.no -= 1
+            return None
+        return expected_hardware
+
+    handler = DeviceHandler(mocked_instance, 'build')
+    handler.no = num_of_failures
+
+    with mock.patch.object(DeviceHandler, "device_is_available", mock_availability):
+        hardware = handler.get_hardware()
+
+    if raise_exception:
+        assert 'dummy message' in caplog.text.lower()
+    else:
+        assert hardware == expected_hardware
+
+TESTDATA_8 = [
+    ('', '', None, None, False, ['generator_cmd']),
+    ('pyocd', 'product', '--dummy-option-1,-do2', [], False, ['--dummy-option-1', '-do2', '--runner', 'pyocd', '--dev-id', 'board_id']),
+    ('openocd', 'STM32 STLink', [], ['--dummy-option3'], False, ['--dummy-option3', '--runner', 'openocd', '--cmd-pre-init', 'hla_serial board_id']),
+    ('openocd', 'STLINK-V3', [], [], True, ['--runner', 'openocd', '--cmd-pre-init', 'hla_serial board_id']),
+    ('openocd', 'EDBG CMSIS-DAP', [], [], False, ['--runner', 'openocd', '--cmd-pre-init', 'cmsis_dap_serial board_id']),
+    ('jlink', 'product', [], [], False, ['--runner', 'jlink', '--tool-opt=-SelectEmuBySN  board_id']),
+    ('stm32cubeprogrammer', 'product', [], [], False, ['--runner', 'stm32cubeprogrammer', '--tool-opt=sn=board_id']),
+]
+
+@pytest.mark.parametrize("runner, product, west_flash_options, runner_options, serial_not_pty, expected", TESTDATA_8)
+def test_devicehandler_handle(mocked_instance, runner, product, west_flash_options, runner_options, serial_not_pty, expected):
```

Comment:
imo, without `handle()` refactoring, this new test might become an additional hindrance and a problem to maintain it with minor changes/fixes.
m.b. we should keep this function not covered with tests now until its refactoring ?

Reply:
As we do not know for know when or even if such a refactoring will come, it might still be a good idea to cover it. The DeviceHandler's `handle` is not so complex that the maintenance would be impossible, its main problem is the length and the amount of responsibilities. Because of that splitting this test Into multiple later shouldn't be that hard.